### PR TITLE
Correctly place "Add column" button and allow using it when summarizing

### LINKED ISSUE #43118: Make plus button less prominent
We've implemented a [new right sidebar with column actions](https://github.com/metabase/metabase/issues/42150) like extract and combine to open which you have to click on the floating plus button. This button was designed as blue but it's [too prominent](https://metaboat.slack.com/archives/C01LQQ2UW03/p1716489446763749).
We decided to update styles a bit: https://www.figma.com/design/Rh7vrNTZX4TPAGC1fDtS55/Extract-and-combine---QB-expressivity-and-column-actions?node-id=2671-3025&t=5AVtVDnjcTXiLRAS-4

### LINKED ISSUE #43173: "Add column" button does nothing when summarizing
### Describe the bug

The "Add column" button (the plus button in the column header) does nothing when a table is the result of summarizing.

image

### To Reproduce

1. Browse data -> Orders
2. Summarize by Created at
3. Switch to tabular view
4. Click the blue `+` button in the column header, it does nothign

### Expected behavior

The button should not be visible since it has no purpose.

### Logs

_No response_

### Information about your Metabase installation

```JSON
{
  "browser-info": {
    "language": "en-US",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:126.0) Gecko/20100101 Firefox/126.0",
    "vendor": ""
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.20.1+1",
    "java.vendor": "Eclipse Adoptium",
    "java.vendor.url": "https://adoptium.net/",
    "java.version": "11.0.20.1",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.20.1+1",
    "os.name": "Mac OS X",
    "os.version": "14.2.1",
    "user.language": "en",
    "user.timezone": "Europe/Amsterdam"
  },
  "metabase-info": {
    "databases": [
      "h2"
    ],
    "hosting-env": "unknown",
    "application-database": "h2",
    "application-database-details": {
      "database": {
        "name": "H2",
        "version": "2.1.214 (2022-06-13)"
      },
      "jdbc-driver": {
        "name": "H2 JDBC Driver",
        "version": "2.1.214 (2022-06-13)"
      }
    },
    "run-mode": "dev",
    "version": {
      "date": "2024-05-23",
      "src_hash": "3408dbf541bb1904a57dcb2cf7022f8e76d2b1ba",
      "tag": "v0.1.6-SNAPSHOT",
      "hash": "40074bd"
    },
    "settings": {
      "report-timezone": null
    }
  }
}
```

### Severity

P3

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by romeovs**
Closes https://github.com/metabase/metabase/issues/43173
Closes https://github.com/metabase/metabase/issues/43118

### Description

- Hide the "Add column" shortcut for questions it does not apply to.
- Make the "Add column" shortcut work for questions with aggregations.
- Make the "Add column" shortcut stick to the last column if there is space.
- Update the button to use the outline variant (https://github.com/metabase/metabase/issues/43118)

### How to verify

1. New question -> Sample Dataset -> Orders
2. On a wide screen the "Add column" button should sit next to the last column, on a smaller screen it should be sticky at in the top right of the header
3. Add a summarization (ie. count by Created At)
4. The "Add column" actions should continue to work.

### Demo

Screenshot 2024-05-28 at 10 35 20

https://github.com/metabase/metabase/assets/1250185/78ee46a3-3f3e-43cf-9f0d-d925d66f160f

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

File: frontend/src/metabase/visualizations/components/TableInteractive/TableInteractive.jsx [hunk 1]
```diff
@@ -1067,9 +1070,15 @@ class TableInteractive extends Component {
...
     const gutterColumn = this.state.showDetailShortcut ? 1 : 0;
-    const shortcutColumn = Number(!question?.isArchived());
     const query = question?.query();
     const info = query && Lib.queryDisplayInfo(query);
+    const hasAggregation = cols.some(column => column.source === "aggregation");
+
+    const shortcutColumn =
+      info?.isEditable &&
+      Number(!question?.isArchived()) &&
+      !hasAggregation &&
```
Comment (romeovs) [inline frontend/src/metabase/visualizations/components/TableInteractive/TableInteractive.jsx:1080]: This seems a bit hacky, but `Lib.expressionableColumns` returns a bunch of columns for the query.

Applying an expression to those columns results in a broken query.

File: e2e/test/scenarios/visualizations-tabular/column-shortcuts/combine-column-shortcut.cy.spec.ts [hunk 2]
```diff
@@ -55,6 +55,26 @@ describeWithSnowplow("scenarios > visualizations > combine shortcut", () => {
       database_id: SAMPLE_DB_ID,
     });
   });
+
+  it("should not show the shortcut button when the table is raw", function () {
```
Comment (uladzimirdev) [inline e2e/test/scenarios/visualizations-tabular/column-shortcuts/combine-column-shortcut.cy.spec.ts:59]: nit: can't it be a unit test?

File: frontend/src/metabase/visualizations/components/TableInteractive/TableInteractive.jsx [hunk 3]
```diff
@@ -1067,9 +1070,15 @@ class TableInteractive extends Component {
...
     const headerHeight = this.props.tableHeaderHeight || HEADER_HEIGHT;
     const gutterColumn = this.state.showDetailShortcut ? 1 : 0;
-    const shortcutColumn = Number(!question?.isArchived());
     const query = question?.query();
     const info = query && Lib.queryDisplayInfo(query);
+    const hasAggregation = cols.some(column => column.source === "aggregation");
+
+    const shortcutColumn =
+      info?.isEditable &&
+      Number(!question?.isArchived()) &&
```
Comment (uladzimirdev) [inline frontend/src/metabase/visualizations/components/TableInteractive/TableInteractive.jsx:1079]: `Number`?

File: frontend/src/metabase/visualizations/components/TableInteractive/TableInteractive.jsx [hunk 4]
```diff
@@ -1321,13 +1331,27 @@ const DetailShortcut = forwardRef((_props, ref) => (
...
 DetailShortcut.displayName = "DetailShortcut";
 
-function ColumnShortcut({ height, onClick, isEditable }) {
-  if (!isEditable) {
+function ColumnShortcut({ height, pageWidth, totalWidth, onClick }) {
+  if (!totalWidth) {
     return null;
   }
 
+  const width = height + 4;
```
Comment (uladzimirdev) [inline frontend/src/metabase/visualizations/components/TableInteractive/TableInteractive.jsx:1339]: could you please name that `4` or leave a comment?

File: e2e/test/scenarios/visualizations-tabular/column-shortcuts/combine-column-shortcut.cy.spec.ts (hunk 2 above)
Comment (romeovs) [reply e2e/test/scenarios/visualizations-tabular/column-shortcuts/combine-column-shortcut.cy.spec.ts:59]: There are currently no unit tests for this component and there is a lot of setup that might need to happen that I'm not sure I'm comfortable enough with to provide.

File: frontend/src/metabase/visualizations/components/TableInteractive/TableInteractive.jsx (hunk 1 above)
Comment (romeovs) [reply frontend/src/metabase/visualizations/components/TableInteractive/TableInteractive.jsx:1080]: @bshepherdson Should this be allow at all?

File: e2e/test/scenarios/visualizations-tabular/column-shortcuts/combine-column-shortcut.cy.spec.ts (hunk 2 above)
Comment (uladzimirdev) [reply e2e/test/scenarios/visualizations-tabular/column-shortcuts/combine-column-shortcut.cy.spec.ts:59]: I mean this test looks exactly like a unit test, but I understand why you did like this

File: frontend/src/metabase/visualizations/components/TableInteractive/TableInteractive.jsx [hunk 5]
```diff
@@ -1321,13 +1331,29 @@ const DetailShortcut = forwardRef((_props, ref) => (
...
-function ColumnShortcut({ height, onClick, isEditable }) {
-  if (!isEditable) {
+const COLUMN_SHORTCUT_PADDING = 4;
+
+function ColumnShortcut({ height, pageWidth, totalWidth, onClick }) {
+  if (!totalWidth) {
     return null;
   }
 
+  const width = height + COLUMN_SHORTCUT_PADDING;
```
Comment (uladzimirdev) [inline frontend/src/metabase/visualizations/components/TableInteractive/TableInteractive.jsx:1341]: I'm curious why do we calculate width based on height? is it kind of design requirement?

File: frontend/src/metabase/visualizations/components/TableInteractive/TableInteractive.jsx (hunk 1 above)
Comment (bshepherdson) [reply frontend/src/metabase/visualizations/components/TableInteractive/TableInteractive.jsx:1080]: See my comments on Slack for more.

I think the right thing here isn't blocking this button, but making it work fine for summarized queries.

File: frontend/src/metabase/visualizations/components/TableInteractive/TableInteractive.jsx (hunk 5 above)
Comment (romeovs) [reply frontend/src/metabase/visualizations/components/TableInteractive/TableInteractive.jsx:1346]: It's square a button with some padding on the left and right.

**Comment by romeovs** [comment]
@uladzimirdev I've updated the PR to reflect the working extractions. Please look again.

**Comment by uladzimirdev** [comment]
@romeovs some tests seems failed

File: frontend/src/metabase/visualizations/components/TableInteractive/TableInteractive.jsx [hunk 6]
```diff
@@ -1067,13 +1074,32 @@ class TableInteractive extends Component {
 
     const headerHeight = this.props.tableHeaderHeight || HEADER_HEIGHT;
     const gutterColumn = this.state.showDetailShortcut ? 1 : 0;
-    const shortcutColumn = Number(!question?.isArchived());
-    const query = question?.query();
-    const info = query && Lib.queryDisplayInfo(query);
+
+    const shortcutActions =
+      question && mode?.clickActions
+        ? mode.clickActions.flatMap(action =>
```
Comment (uladzimirdev) [inline frontend/src/metabase/visualizations/components/TableInteractive/TableInteractive.jsx:1077]: `question && mode?.clickActions?.flatMap`

Comment (uladzimirdev) [reply frontend/src/metabase/visualizations/components/TableInteractive/TableInteractive.jsx:1077]: and do we really need to iterate over all click actions just to understand if it's greater than 0?

**Review by uladzimirdev — APPROVED** [review]
From code perspective LGTM, but I'd wait for someone from @metabase/core-frontend-dashviz to verify TableInteractive changes

File: frontend/src/metabase/visualizations/click-actions/actions/CombineColumnsAction/CombineColumnsAction.tsx [hunk 7]
```diff
@@ -9,14 +9,18 @@ import type { ClickActionPopoverProps } from "metabase/visualizations/types/clic
...
+
+  const { isEditable } = Lib.queryDisplayInfo(query);
+  const isExpressionable =
+    Lib.expressionableColumns(query, stageIndex).length > 0;
 
   if (
     !clicked ||
     clicked.value !== undefined ||
     !clicked.columnShortcuts ||
-    clicked.extraData?.isRawTable ||
```
Comment (romeovs) [inline frontend/src/metabase/visualizations/click-actions/actions/CombineColumnsAction/CombineColumnsAction.tsx:18]: I'm not 100% sure what `isRawTable` means, but I think the other heuristics are enough.

**Review by alxnddr — COMMENTED** [review]
Looks good! I noticed one issue with the functionality, please let me know what you think.

1. Orders -> Summarize Count of rows by Created At
2. Make it a Line chart
3. At the bottom, toggle data view which shows the table
4. Add a new column from this view
5. The viz type has changed from Line to Table

**Comment by romeovs** [comment]
@alxnddr that seems like an unrelated issue, I've created an issue for it, but let's not let it block this PR: https://github.com/metabase/metabase/issues/43294

**Pull request merged**